source/language/openpulse.rst

```diff
@@ -0,0 +1,316 @@
+for the channel will be used instead, as if ``frameof`` was called for that
+channel.
+
+.. code-block:: none
+
+   // Play a 3 sample pulse on qubit 0
+   play transmit(%0), [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
+
+   // Play a gaussian on qubit 0 using qubit 1's frame
+   frame f1 = frameof(transmit(%1));
+   play transmit(%0), gaussian(...), f1;
+
+Capture Instruction
+-------------------
+
+Acquisition is scheduled by a ``capture`` instruction. This is a special
+``kernel`` function which is specified by a hardware vendor. The measurement
+process is difficult to describe generically due to the wide variety of
+hardware and measurement methods.
+
+The only required parameter is an ``channel``, ie. a channel returned from
+calling the ``receive`` function.
+
+The following are possible parameters that might be included:
+
+- A "duration" of type ``length``, if it cannot be inferred from other parameters
+- A "filter", which is dot product-ed with the measured IQ the distill the
+  result into a single IQ value
+- A "tag", which could be used to identify which branch of an if statement was
+  traversed
+
+Again it is up to the hardware vendor to determine the parameters and write a
+kernel definition at the top-level, such as:
+
+.. code-block:: none
+
+   // Minimum requirement
+   kernel capture(channel output) -> complex[32];
+
+   // A capture command with more features
+   kernel capture(channel output, pulse filter) -> complex[32];
+
+The return type of a ``capture`` command varies. It could be a raw trace, ie a
+list of samples taken over a short period of time. It could be some averaged IQ
+value. It could be a classified bit. Or it could even have no return value,
+pushing the results into some buffer which is then accessed outside the program.
+
+Timing
+------
+
+Each channel maintains its own "clock". When a pulse is played the clock for 
+that channel advances by the length of the pulse. The same is true
+for output channels. Pulses on a single channel cannot be played simultaneously,
+although pulses on multiple channels for the same qubit can.
+
+For channels, everything behaves analogous to qubits in the 
+`Delays <delays.html>`_ section of this specification. There are however some
+small differences.
+
+The ``delay`` instruction may take a channel instead of a qubit. If a ``delay``
+instruction is applied to the qubit, this is the same as applying the delay to
+all channels on the qubit simultaneously.
+
+The ``barrier`` instruction on a qubit implies a barrier on all channels defined
+for that qubit. A barrier instruction will advance the clocks on all channels of
+the qubit to the channel with the highest clock.
+
+``defcal`` blocks have an implicit barrier on every qubit argument, meaning
+that clocks are guaranteed to be aligned at the start of the ``defcal`` block.
+These blocks also need to have a well-defined length, similar to the ``boxas``
+block.
```

Comment:
How do barriers on qubits get converted into barriers on channels? I am again thinking about the case of multiplexed measurement where i want:
```
measure %0 -> c[0];
measure %1 -> c[1];
```
where qubits 0 and 1 are on a multiplexed feedline to still issue simultaneously.

Reply:
That's a great question. The solution described in the current diff says something like "a barrier on a qubit applies to all channels for that qubit", which fails to provide a way to do multiplexed measurement.

There might even be a bigger problem here: if you're doing multiplexed readout of multiple qubits at once, how do you communicate that back to the individual measure instructions?
ie. this is really doing something like `measure %0 -> c[0], %1 -> c[1];`

Reply:
I'm not looking for a single instruction that does multiplexed measurement, but rather I want to make sure we build the grammar and semantics so that at the gate/operation-level I can issue independent `measure` instructions and have that be compatible with a pulse-level implementation where the signals share a channel because the physical implementation is multiplexed.

Reply:
+1, another good argument for implementing timing on frames rather than channels.